sqlstats: add batching to flush operation

dhartunian opened the pull request:
Previously, we would process each persisted stat row insert as a separate transaction. This incurred lots of transaction teardown/setup costs over the life of the flush operation.

This change modifies the flush to batch the inserts in groups of 100 (arbitrarily chosen, for now). The batches are iterated over using a small struct that constructs the batch slices from the list of stats. The rest of the operation remains as-is except the SQL transaction is moved upwards in the callstack to contain all the elements in a single batch.

Resolves: #139143

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139608)
<!-- Reviewable:end -->

dhartunian committed:
sqlstats: remove context argument to visitor functions

The context argument to the transaction and statement visitor
functions was unused. This makes future modifications of the flush
simpler.

The only valid use was to pass a `ctx` into a logging statement, but
this change preserves the same `ctx` from the top-level caller.

Release note: None

dhartunian committed:
sqlstats: add batching to flush operation

Previously, we would process each persisted stat row insert as a
separate transaction. This incurred lots of transaction teardown/setup
costs over the life of the flush operation.

This change modifies the flush to batch the inserts in batches sized
via a cluster setting (`sql.stats.flush.batch_size`). The batches are
iterated over using a small struct that constructs the batch slices
from the list of stats. The rest of the operation remains as-is except
the SQL transaction is moved upwards in the callstack to contain all
the elements in a single batch.

Resolves: #139143

Release note: None

dhartunian commented:
@kyle-a-wong merged a different batching PR here: https://github.com/cockroachdb/cockroach/pull/140673
